src/FastPriorityQueue.php

```diff
@@ -0,0 +1,298 @@
+     * order of insertion
+     *
+     * @return mixed
+     */
+    public function extract()
+    {
+        if (!$this->valid()) {
+            return false;
+        }
+        $value = $this->current();
+        $this->nextAndRemove();
+        return $value;
+    }
+
+    /**
+     * Get the total number of elements in the queue
+     *
+     * @return integer
+     */
+    public function count()
+    {
+        return $this->tot;
+    }
+
+    /**
+     * Get the current element in the queue
+     *
+     * @return mixed
+     */
+    public function current()
+    {
+        switch ($this->extractFlag) {
+            case self::EXTR_DATA:
+                return current($this->values[$this->max]);
+            case self::EXTR_PRIORITY:
+                return $this->max;
+            case self::EXTR_BOTH:
+                return [
+                    'data' => current($this->values[$this->max]),
+                    'priority' => $this->max
+                ];
+        }
+    }
+
+    /**
+     * Get the index of the current element in the queue
+     *
+     * @return integer
+     */
+    public function key()
+    {
+        return $this->index;
+    }
+
+    /**
+     * Set the iterator pointer to the next element in the queue
+     * removing the previous element
+     */
+    protected function nextAndRemove()
+    {
+        if (false === next($this->values[$this->max])) {
+            unset($this->priorities[$this->max]);
+            unset($this->values[$this->max]);
+            $this->max = empty($this->priorities) ? 0 : max($this->priorities);
+        }
+        ++$this->index;
+        --$this->tot;
+    }
+
+    /**
+     * Set the iterator pointer to the next element in the queue
+     * without removing the previous element
+     */
+    public function next()
+    {
+        if (false === next($this->values[$this->max])) {
+            unset($this->subPriorities[$this->max]);
+            $this->max = empty($this->subPriorities) ? 0 : max($this->subPriorities);
+        }
+        ++$this->index;
+        --$this->tot;
+    }
+
+    /**
+     * Check if the current iterator is valid
+     *
+     * @return boolean
+     */
+    public function valid()
+    {
+        return isset($this->values[$this->max]);
+    }
+
+    /**
+     * Rewind the current iterator
+     */
+    public function rewind()
+    {
+        $this->subPriorities = $this->priorities;
+    }
+
+    /**
+     * Serialize to an array
+     *
+     * Array will be priority => data pairs
+     *
+     * @return array
+     */
+    public function toArray()
+    {
+        $array = [];
+        foreach (clone $this as $item) {
+            $array[] = $item;
+        }
+        return $array;
+    }
+
+    /**
+     * Serialize
+     *
+     * @return string
+     */
+    public function serialize()
+    {
+        $clone = clone $this;
+        $clone->setExtractFlags(self::EXTR_BOTH);
```

Comment:
To ensure that when the clone is unserialized, it's in the same state as the original.